Docs: document editing cortex/loki recording rules, namespaces and groups (#38547)
grafana/grafana: docs/sources/alerting/unified-alerting/alerting-rules/create-cortex-loki-managed-recording-rule.md

+++
title = "Create Cortex or Loki managed recording rule"
description = "Create Cortex or Loki managed recording rule"
keywords = ["grafana", "alerting", "guide", "rules", "recording rules", "create"]
weight = 400
+++

> # Create a Cortex or Loki managed recording rule

Reviewer: Do we actually call Metrics and Logging by the names Cortex and Loki in Cloud docs as well as OSS Docs?

Reviewer 2: > Do we actually call Metrics and Logging by the names Cortex and Loki in Cloud docs as well as OSS Docs?

@richardqlam, can you shed more light on this. Will help us working on OSS docs as well. Thank you.

Reviewer 3: Talked to @Eve832 about this but in general, we like to talk about Cloud metrics and Cloud logs in a broad sense. It's okay to use the specific OSS terminology where it's relevant though, because sometimes it's necessary.  

For example, let's first consider metrics. It's not totally relevant to users that we use Cortex as our metrics backend in Grafana Cloud. What matters, though, is that "Grafana Cloud metrics are compatible with Prometheus and Graphite metrics". 

But when you look at Grafana Cloud logs, the story is a bit different. It _requires_ you to understand the kind of logging system we're using (Loki), because the logs are structured completely different than a system like Elastic and fit a totally different use case.

Grafana allows you manage recording rules for an external Cortex or Loki instance.

Recording rules allow you to precompute frequently needed or computationally expensive expressions and save their result as a new set of time series. Querying the precomputed result will then often be much faster than executing the original expression every time it is needed. This is especially useful for dashboards, which need to query the same expression repeatedly every time they refresh.

> Recording and alerting rules exist in a rule group. Rules within a group are run sequentially at a regular interval, with the same evaluation time. The names of recording rules must be valid [metric names](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels). 

Reviewer 2: ```suggestion
Recording and alerting rules exist in a rule group and they are run sequentially at a regular interval with the same evaluation time. 
```

Author: you think the caveat about reacording rule names being valid metric names is not necesary?

Author: nvm, noticed it was just moved

> In order for both Cortex and Loki data sources to work with Grafana 8.0 alerting, enable the ruler API by configuring their respective services. The`local` rule storage type, default for Loki, only supports viewing of rules. If you want to edit rules, then configure one of the other rule storage types. When configuring a Grafana Prometheus data source to point to Cortex, use the legacy `/api/prom` prefix, not `/prometheus`. Only single-binary mode is currently supported, and it is not possible to provide a separate URL for the ruler API.

Author: thanks! I didn't know how to do that

## Add or edit a Cortex or Loki managed recording rule

1. In the Grafana menu hover your cursor over the Alerting (bell) icon.
1. To create a new recording rule, click **New alert rule**. To edit an existing rule, expand one of the recording rules in the **Cortex / Loki** section and click **Edit**.
1. Click on the **Alert type** drop down and select **Cortex / Loki managed recording rule**.
1. Fill out the rest of the fields. Descriptions are listed below in [Recording rule fields](#recording-rule-fields).

> 1. When you have finished writing your rule, click **Save** in the upper right corner to save the rule, or **Save and exit** to save and exit rule editing.

Author: *Save* and *Save and exit* buttons exist both when creating new and when editing, not *Save* for creating and *Save and exit* for editing
